docs/references/message-execution-properties.mdx

```diff
@@ -51,12 +51,18 @@ Property 3 provides a guarantee on the execution order of messages on a target c
+There are many reasons why a call might not be delivered to the destination canister. Some of them are under the control of the canister developers, such as the destination having sufficient cycles to process the call. Others are not; the Internet Computer may decide to fail the call if it's under high load.
 
-For more details, refer to the Interface Specification [section on ordering guarantees](/docs/current/references/ic-interface-spec#ordering_guarantees) and the section on [abstract behavior](/docs/current/references/ic-interface-spec#abstract-behavior) which defines message execution in more detail.
\ No newline at end of file
+- **Property 7**: Every inter-canister call is guaranteed to receive a response, either from the canister, or synthetically produced by the protocol.
+
+However, a malicious destination canister could choose to delay the response for arbitrarily long if it is willing to put in the required cycles. Also, the response does not have to be successful, but can also be a reject response. The reject may come from the called canister, but it may also be generated by ICP. Such protocol-generated rejects can occur at any time before the call reaches the callee-canister, as well as once the call does reach the callee-canister, for example if the callee-canister traps while processing the call. Thus, it's important that the calling canister handles reject responses as well. A reject response means that the message hasn't been successfully processed by the receiver but doesn't guarantee that the receiver's state wasn't changed.
+
+- **Property 8**: If the calling canister made a guaranteed response (as opposed to a best-effort response) inter-canister call, and the call reaches the callee-canister which then produces a reply or reject response, the protocol guarantees that the callee-canister's generated reply or reject response gets back to the caller-canister.
```

Comment:
I don't think this is broad enough. I would put it in terms of "for guaranteed response calls there is always exactly one globally unique response to a call (whether from the callee or from the protocol)".

Or even "there is always one single globally unique response for every canister call (originating from either the callee or the protocol); if the call is a guaranteed response call, then this unique response is guaranteed to be delivered to the caller; if it was a best-effort call, then a `SYS_UNKNOWN` response may be delivered instead (and potentially before)". Or something along those lines.

Reply:
Overall, I'm not sure how useful it is to talk about the "real response generated by the system" in case of `SYS_UNKNOWN`, as this real response is not something observable by the user. What is observable is that the callee produced a response but it didn't get delivered. So we can introduce the internal requests and responses (including system-generated responses) if it's the simplest mental model for the user of what's happening, but if there's a simple way to explain everything using just what's observable, I'd prefer that.

Not entirely related, but my original idea was to have a property that only applies to guaranteed response calls, and we could just say "best-effort responses lack this property". But thinking about it, it's probably useful to have a separate property about losing responses that we can refer to.

Reply:
> The system itself may also produce a reject response without making the call at all and this is also guaranteed to be delivered to the caller. And, importantly, there is always exactly one outcome for every call.

This is covered by Property 7. This property just says that, for guaranteed response calls, if the caller actually produces a response, the user will get that response.

> The difference in the best-effort response case is that the system may at any time, including before the call completes or even begins execution, produce a SYS_UNKNOWN reject response 

Of course it's also true that for guaranteed response calls, if the system produces *some* response, only *that* response will be delivered to the user, which is not true for best-effort responses, and what you write is an entirely accurate description how the system works. What I'm saying is that in general, the user doesn't *need* to know how the system works. We just have to give them the simplest correct description of the behavior that they can observe. It might well be that the simplest explanation is to tell them how the system really works. I tried providing a simple description without going into the system details but it's quite possible that I failed. I hope my intention is clear now - if it was clear all along and you feel that it's too complicated or not precise enough, I will switch to what you proposed, otherwise take another look and LMK what you think.

Reply:
Maybe what's missing is to say something like "the call is only deliver at most once" in Property 5?

Reply:
I just found it unnecessarily specific. IMHO there is no need to qualify this with "if the call is delivered to the callee and the callee responds without trapping".

But yeah, if you want to describe strictly the observable properties of the system, then this is more accurate. So it's entirely your call.